### ISSUE TITLE
Issues with duplicate reporting in CSV import

### ISSUE THREAD

**Issue by dae** [ankitects/anki#2581]
I found some issues with CSV import logging:

1. If the file contains multiple duplicate notes, notes can be reported more than once in the duplicate queue.

<img src="https://github.com/ankitects/anki/assets/41397710/6a7136c2-ad92-4083-8f75-4616e77513bc" width=600>

2. If the collection itself has duplicates that exist in the file, all collection duplicates are reported, creating another mismatch in the counts. (screenshot taken after applying a quick fix for the first issue)

<img src="https://github.com/ankitects/anki/assets/41397710/3e699728-3281-49f9-93b8-1374e57f2ea3" width=600>

Test file:
[importtest.txt](https://github.com/ankitects/anki/files/12061288/importtest.txt)

_Originally posted by @abdnh in https://github.com/ankitects/anki/issues/2568#issuecomment-1636745967_

**Comment by dae** [comment]
These looks like corner cases that won't pop up *too* frequently. What do you think @RumovZ, are these worth addressing with an extra hashset or similar, or leaving alone for now?

**Comment by RumovZ** [comment]
I think issue 2 is related to the caveat in `import_export.proto`:
```protobuf
    // Usually the sum of all queues, but may be lower if multiple duplicates
    // have been updated with the same note.
    uint32 found_notes = 10;
```
The import log rework seems like a compelling reason to rectify this. Here's one idea:
Change "X were updated to the latest version" to something along the lines of "X were used to update existing ones", optionally adding "(Y affected)" or similar. Then in the code, instead of logging every duplicate of a given incoming note, only one would be recorded. The strategy to choose which one, might look like this:
```py
if any(updated): log(updated[0])
elif any(duplicate): log(duplicate[0])
elif any(conflicting): log(conflicting[0])
else: assert_never()
```
That should be a fairly simple change.

As for issue 1, I cannot reproduce it and don't see why it would happen. I'm assuming you've actually observed issue 2 here as well, but please correct me if I'm wrong, @abdnh.

**Comment by dae** [comment]
Presumably the main confusion here is when the displayed counts don't match the count from the file, so showing a count based on the imported file instead of the target notes (eg "Of those, 300 matched existing notes", ".. were used to update existing notes", etc) does sound good. If we can get away without including "Y affected" that might be better, as the extra info does make it harder to read/doesn't help in the common case.

**Comment by abdnh** [comment]
> As for issue 1, I cannot reproduce it and don't see why it would happen. I'm assuming you've actually observed issue 2 here as well

You should be able to reproduce it by first importing the file, then importing it again with DupeResolution::Update set.

I found about issue 2 after applying the following quick fix for issue 1:
```diff
diff --git a/rslib/src/import_export/text/import.rs b/rslib/src/import_export/text/import.rs
index 4b64e49a5..f0bce9d9f 100644
--- a/rslib/src/import_export/text/import.rs
+++ b/rslib/src/import_export/text/import.rs
@@ -404,7 +404,10 @@ impl<'a> Context<'a> {
             self.add_cards(&mut cards, &note, ctx.deck_id, ctx.notetype.clone(), log)?;

             if dupe.identical {
-                log.duplicate.push(dupe.note.into_log_note());
+                let log_note = dupe.note.into_log_note();
+                if !log.duplicate.contains(&log_note) {
+                    log.duplicate.push(log_note);
+                }
             } else if dupe.first_field_match {
                 log.first_field_match.push(note.into_log_note());
             } else {
```

**Comment by abdnh** [comment]
@RumovZ You should actually see the following result for issue 1.
![image](https://github.com/ankitects/anki/assets/41397710/f4caf976-6641-4167-940d-d1854d464f98)

(My first screenshot was probably taken with some other cards in the collection or a different file; I don't exactly remember).

**Comment by RumovZ** [comment]
> You should be able to reproduce it by first importing the file, then importing it again with DupeResolution::Update set.

It says above that issue 1 is about duplicates in the *file*. If that was the case, I wouldn't need to import the file twice to observe it, would I? On the second import, my collection will have duplicates, so the number can be explained by issue 2. (300 notes in the file \* 150 dupes each in the collection = 45,000.)

**Comment by abdnh** [comment]
I think you're right. It makes sense to me now after taking another look at the code.

**Comment by RumovZ** [comment]
Resolved with #2568 / https://github.com/abdnh/anki/pull/9.
